central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -523,7 +535,120 @@ func (ds *datastoreImpl) RemovePlopsByPod(ctx context.Context, id string) error
 	return storeErr
 }
+
+// PruneOrphanedPLOPs prunes old closed PLOPs and those without deployments or pods
+func (ds *datastoreImpl) PruneOrphanedPLOPs(ctx context.Context, orphanWindow time.Duration) int64 {
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+
+	query := fmt.Sprintf(pruneOrphanedPLOPs, int(orphanWindow.Minutes()))
+	commandTag, err := ds.pool.Exec(ctx, query)
+	if err != nil {
+		log.Errorf("failed to prune PLOP: %v", err)
+	}
+
+	// Delete processes listening on ports orphaned due to missing deployments
+	if _, err := ds.pool.Exec(ctx, deleteOrphanedPLOPDeployments); err != nil {
+		log.Errorf("failed to prune process listening on ports by deployment: %v", err)
+	}
+
+	// Delete processes listening on ports orphaned due to missing pods.
+	if _, err := ds.pool.Exec(ctx, deleteOrphanedPLOPPodsWithPodUID); err != nil {
+		log.Errorf("failed to prune process listening on ports by pods: %v", err)
+	}
+
+	return commandTag.RowsAffected()
+}
+
+// PruneOrphanedPLOPsByProcessIndicators prunes PLOPs that match process indicators without pods
+func (ds *datastoreImpl) PruneOrphanedPLOPsByProcessIndicators(ctx context.Context, orphanWindow time.Duration) {
+	// TODO(ROX-22443): Once it is guaranteed that all listening endpoints have PodUIDs, remove this function
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+
+	// Delete processes listening on ports orphaned because process indicators are orphaned due to
+	// missing deployments
+	query := fmt.Sprintf(deleteOrphanedPLOPDeploymentsAndPI, int(orphanWindow.Minutes()))
+	if _, err := ds.pool.Exec(ctx, query); err != nil {
+		log.Errorf("failed to prune process listening on ports by deployment: %v", err)
+	}
+
+	// Delete processes listening on ports orphaned because process indicators are orphaned due to
+	// missing pods.
+	query = fmt.Sprintf(deleteOrphanedPLOPPods, int(orphanWindow.Minutes()))
+	if _, err := ds.pool.Exec(ctx, query); err != nil {
+		log.Errorf("failed to prune process listening on ports by pods: %v", err)
+	}
+}
+
+func (ds *datastoreImpl) readRowsToFindPLOPsWithNoProcessInformation(rows pgx.Rows) ([]string, error) {
+	var ids []string
+
+	for rows.Next() {
+		var serialized []byte
+
+		if err := rows.Scan(&serialized); err != nil {
+			return nil, pgutils.ErrNilIfNoRows(err)
+		}
+
+		var msg storage.ProcessListeningOnPortStorage
+		if err := proto.Unmarshal(serialized, &msg); err != nil {
+			return nil, err
+		}
+
+		process := msg.GetProcess()
+
+		if process == nil {
+			ids = append(ids, msg.GetId())
+		}
+	}
+	return ids, nil
+}
+
+// First gets PLOPs with no matching process indicators and then checks the
+// serialized data to check for process information.
+func (ds *datastoreImpl) getPLOPsToDelete(ctx context.Context) ([]string, error) {
+	rows, err := ds.pool.Query(ctx, getPotentiallyOrphanedPLOPs)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer rows.Close()
+
+	plopsToDelete, err := ds.readRowsToFindPLOPsWithNoProcessInformation(rows)
```

Comment:
just return the function instead of having intermediate vars

Reply:
Done